src/web/tokenserver.rs

```diff
@@ -1,26 +1,165 @@
 use actix_web::error::BlockingError;
 use actix_web::web::block;
 use actix_web::HttpResponse;
+use actix_web_httpauth::extractors::bearer::BearerAuth;
 
 use futures::future::{Future, TryFutureExt};
 
-use crate::error::ApiError;
-use crate::web::extractors::TokenServerRequest;
+use crate::error::{ApiError, ApiErrorKind};
 
-pub struct TokenServerResult {}
+use diesel::mysql::MysqlConnection;
+use diesel::prelude::*;
+use diesel::sql_types::*;
+use diesel::RunQueryDsl;
+use std::env;
+
+use jsonwebtoken::{decode, Algorithm, DecodingKey, Validation};
+use pyo3::prelude::*;
+use pyo3::types::IntoPyDict;
+
+#[derive(Debug)]
+enum MyError {
+    EnvError(env::VarError),
+}
+
+impl From<env::VarError> for MyError {
+    fn from(error: env::VarError) -> Self {
+        MyError::EnvError(error)
+    }
+}
+
+#[derive(Debug, QueryableByName)]
+struct TokenserverUser {
+    #[sql_type = "Bigint"]
+    uid: i64,
+    // This is no longer used. Was for making more than just sync tokens.
+    #[sql_type = "Text"]
+    pattern: String,
+    #[sql_type = "Text"]
+    email: String,
+    #[sql_type = "Bigint"]
+    generation: i64,
+    #[sql_type = "Text"]
+    client_state: String,
+    #[sql_type = "Bigint"]
+    created_at: i64,
+    #[sql_type = "Nullable<Bigint>"]
+    replaced_at: Option<i64>,
+    #[sql_type = "Text"]
+    node: String,
+    #[sql_type = "Nullable<Bigint>"]
+    keys_changed_at: Option<i64>,
+}
+
+#[derive(serde::Serialize)]
+pub struct TokenServerResult {
+    id: String,
+    key: String,
+    uid: String,
+    api_endpoint: String,
+    duration: String,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct Claims {
+    pub sub: String,
+    pub iat: i64,
+    pub exp: i64,
+}
 
 pub fn get(
-    request: TokenServerRequest,
+    auth: BearerAuth,
 ) -> impl Future<Output = Result<HttpResponse, BlockingError<ApiError>>> {
-    block(move || get_sync(request).map_err(Into::into)).map_ok(move |_result| {
-        // TODO turn _result into a json response.
+    block(move || get_sync(&auth).map_err(Into::into)).map_ok(move |result| {
         HttpResponse::Ok()
             .content_type("application/json")
-            .body("{}")
+            .body(serde_json::to_string(&result).unwrap())
     })
 }
 
-pub fn get_sync(_request: TokenServerRequest) -> Result<TokenServerResult, ApiError> {
-    // TODO Perform any blocking calls needed to respond to the tokenserver request.
-    Ok(TokenServerResult {})
+pub fn get_sync(auth: &BearerAuth) -> Result<TokenServerResult, ApiError> {
+    // the public rsa components come from
+    // https://oauth.accounts.firefox.com/v1/jwks
+    // TODO we should fetch it from an environment var instead of hardcoding it.
```

Comment:
yes. this.

Reply:
I'm going to leave this for now and do it in bug #898